### PR TITLE
Download unknown but announced forks

### PR DESCRIPTION
**Pull request by rphmeier**
related: #1539 #1593

### CODE REVIEW

File: core/finality-grandpa/src/communication.rs [hunk 1]
```diff
@@ -17,6 +17,7 @@
 //! Incoming message streams that verify signatures, and outgoing message streams
 //! that sign or re-shape.
 
+use std::collections::HashMap;
```
Comment (bkchr) [inline core/finality-grandpa/src/communication.rs]: This is not used?

File: core/network/src/protocol.rs [hunk 2]
```diff
@@ -705,12 +706,11 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 		let mut peers = self.context_data.peers.write();
 		let hash = header.hash();
 		for (who, ref mut peer) in peers.iter_mut() {
-			if peer.known_blocks.insert(hash.clone()) {
-				trace!(target: "sync", "Reannouncing block {:?} to {}", hash, who);
-				self.send_message(io, *who, GenericMessage::BlockAnnounce(message::BlockAnnounce {
-					header: header.clone()
-				}));
-			}
+			peer.known_blocks.insert(hash.clone());
```
Comment (bkchr) [inline core/network/src/protocol.rs]: Why do we need to clone this hash? Couldn't we just put the trace one line above?

File: core/network/src/sync.rs [hunk 3]
```diff
@@ -40,6 +40,9 @@ const MAJOR_SYNC_BLOCKS: usize = 5;
 const JUSTIFICATION_RETRY_WAIT: Duration = Duration::from_secs(10);
 // Number of recently announced blocks to track for each peer.
 const ANNOUNCE_HISTORY_SIZE: usize = 64;
+// Max number of blocks to download for unknown forks.
+// TODO: this should take finality into account. See https://github.com/paritytech/substrate/issues/1606
+const MAX_UNKOWN_FORK_DOWNLOAD_LEN: u32 = 32;
```
Comment (andresilva) [inline core/network/src/sync.rs]:
```suggestion
const MAX_UNKNOWN_FORK_DOWNLOAD_LEN: u32 = 32;
```

File: core/finality-grandpa/src/communication.rs (hunk 1 above)
Comment (rphmeier) [reply core/finality-grandpa/src/communication.rs]: it is used, a merge earlier caused the build to break

Comment (arkpar) [reply core/finality-grandpa/src/communication.rs]: It is used. This branch did not compile before this change and I guess master ir broken too.

File: core/network/src/protocol.rs (hunk 2 above)
Comment (rphmeier) [reply core/network/src/protocol.rs]: it definitely needs to be cloned because this is in a loop. the trace is not relevant to that

**Pull request merged**